// @angular-eslint/eslint-plugin-template/dist/rules/no-duplicate-attributes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RULE_NAME = void 0;
const utils_1 = require("@angular-eslint/utils");
const create_eslint_rule_1 = require("../utils/create-eslint-rule");
const get_original_attribute_name_1 = require("../utils/get-original-attribute-name");
exports.RULE_NAME = 'no-duplicate-attributes';
const DEFAULT_OPTIONS = {
    allowTwoWayDataBinding: true,
    ignore: [],
};
exports.default = (0, create_eslint_rule_1.createESLintRule)({
    name: exports.RULE_NAME,
    meta: {
        type: 'problem',
        docs: {
            description: 'Ensures that there are no duplicate input properties or output event listeners',
            recommended: false,
        },
        hasSuggestions: true,
        schema: [
            {
                type: 'object',
                properties: {
                    allowTwoWayDataBinding: {
                        type: 'boolean',
                        default: DEFAULT_OPTIONS.allowTwoWayDataBinding,
                        description: `Whether or not two-way data binding is allowed as an exception to the rule.`,
                    },
                    ignore: {
                        type: 'array',
                        items: { type: 'string' },
                        uniqueItems: true,
                        default: DEFAULT_OPTIONS.ignore,
                        description: `Input or output properties for which duplicate presence is allowed as an exception to the rule.`,
                    },
                },
                additionalProperties: false,
            },
        ],
        messages: {
            noDuplicateAttributes: 'Duplicate attribute `{{attributeName}}`',
            suggestRemoveAttribute: 'Remove attribute `{{attributeName}}`',
        },
    },
    defaultOptions: [DEFAULT_OPTIONS],
    create(context, [{ allowTwoWayDataBinding, ignore }]) {
        const parserServices = (0, utils_1.getTemplateParserServices)(context);
        return {
            Element$1({ inputs, outputs, attributes }) {
                const duplicateInputsAndAttributes = findDuplicates([
                    ...inputs,
                    ...attributes,
                ]);
                const filteredOutputs = allowTwoWayDataBinding
                    ? outputs.filter((output) => {
                        return !inputs.some((input) => input.sourceSpan.start === output.sourceSpan.start &&
                            input.sourceSpan.end === output.sourceSpan.end);
                    })
                    : outputs;
                const duplicateOutputs = findDuplicates(filteredOutputs);
                const allDuplicates = [
                    ...duplicateInputsAndAttributes,
                    ...duplicateOutputs,
                ];
                const filteredDuplicates = ignore && ignore.length > 0
                    ? allDuplicates.filter((duplicate) => !ignore.includes((0, get_original_attribute_name_1.getOriginalAttributeName)(duplicate)))
                    : allDuplicates;
                filteredDuplicates.forEach((duplicate) => {
                    const loc = parserServices.convertNodeSourceSpanToLoc(duplicate.sourceSpan);
                    const data = {
                        attributeName: (0, get_original_attribute_name_1.getOriginalAttributeName)(duplicate),
                    };
                    context.report({
                        loc,
                        messageId: 'noDuplicateAttributes',
                        data,
                        suggest: [
                            {
                                messageId: 'suggestRemoveAttribute',
                                fix: (fixer) => fixer.removeRange([loc.start.column, loc.end.column + 1]),
                                data,
                            },
                        ],
                    });
                });
            },
        };
    },
});
function findDuplicates(elements) {
    return elements.filter((element) => {
        return elements.some((otherElement) => otherElement !== element &&
            (0, get_original_attribute_name_1.getOriginalAttributeName)(otherElement) ===
                (0, get_original_attribute_name_1.getOriginalAttributeName)(element));
    });
}
