// nx/src/command-line/show/command-object.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.yargsShowCommand = void 0;
const tslib_1 = require("tslib");
const shared_options_1 = require("../yargs-utils/shared-options");
exports.yargsShowCommand = {
    command: 'show',
    describe: 'Show information about the workspace (e.g., list of projects)',
    builder: (yargs) => yargs
        .command(showProjectsCommand)
        .demandCommand()
        .example('$0 show projects', 'Show a list of all projects in the workspace')
        .example('$0 show projects --affected', 'Show affected projects in the workspace')
        .example('$0 show projects --affected --exclude *-e2e', 'Show affected projects in the workspace, excluding end-to-end projects'),
    handler: (args) => tslib_1.__awaiter(void 0, void 0, void 0, function* () {
        // Noop, yargs will error if not in a subcommand.
    }),
};
const showProjectsCommand = {
    command: 'projects',
    describe: 'Show a list of projects in the workspace',
    builder: (yargs) => (0, shared_options_1.withAffectedOptions)(yargs)
        .option('affected', {
        type: 'boolean',
        description: 'Show only affected projects',
    })
        .implies('untracked', 'affected')
        .implies('uncommitted', 'affected')
        .implies('files', 'affected')
        .implies('base', 'affected')
        .implies('head', 'affected'),
    handler: (args) => Promise.resolve().then(() => require('./show')).then((m) => m.showProjectsHandler(args)),
};
